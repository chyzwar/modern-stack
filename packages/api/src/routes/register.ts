import { RouteShorthandOptions } from "fastify";
import { FastifyServer } from "../types/Server";

const register = async(server: FastifyServer): Promise<void> => {
  const route: RouteShorthandOptions = {
    schema: {
      response: {
        200: {
          type: 'object',
          properties: {
            hello: { type: 'string' }
          }
        }
      }
    },
    handler (request, reply) {
      reply.send({ hello: 'world' })
    }
  }

  server.get("/register", route);
}



export default register;