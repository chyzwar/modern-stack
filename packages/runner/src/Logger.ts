import pc from "picocolors"

class Logger {
  public static calls: unknown[][];
 
  private readonly taskName: string;

  public constructor(taskName: string) {
    this.taskName = taskName;
  }

  public time(label: string): void {
    console.time(pc.green(`[${this.taskName}] ${label}`));
  }

  public timeEnd(label: string): void {
    console.timeEnd(pc.green(`[${this.taskName}] ${label}`));
  }

  public log(message = ""): void {
    console.log(`[${pc.green(this.taskName)}] ${message}`);
  }

  public warn(message = "", meta: unknown = ""): void {
    console.warn(pc.yellow(`[${this.taskName}] ${message}`), meta);
  }

  public error(message = "", meta: unknown = ""): void {
    console.log(`${pc.red(`[${this.taskName}]`)} ${message}`, meta);
  }

  public info(message = ""): void {
    console.log(`${pc.green(`[${this.taskName}]`)} ${message}`);
  }
}


export default Logger;