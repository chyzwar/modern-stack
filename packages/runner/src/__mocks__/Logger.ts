class Logger {
  public static calls: unknown[][] = [];

  public info(...args: unknown[]): void {
    Logger.calls.push(args);
  }

  public time(...args: unknown[]): void {
    Logger.calls.push(args);
  }

  public timeEnd(...args: unknown[]): void {
    Logger.calls.push(args);
  }

  public error(...args: unknown[]): void {
    Logger.calls.push(args);
  }
} 

export default Logger;