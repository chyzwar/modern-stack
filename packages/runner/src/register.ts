import type TaskFunction from "./TaskFunction.js";

const register = new Map<string, TaskFunction>();

export default register;