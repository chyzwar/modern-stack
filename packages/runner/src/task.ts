import type TaskFunction from "./TaskFunction.js";
import register from "./register.js";
import Logger from "./Logger.js";

/**
 * Register task
 */
function task(taskName: string, taskFunction: TaskFunction): void {
  async function innerTask(): Promise<void> {
    const logger = new Logger(taskName);
    logger.info("Started task");

    try {
      logger.time("Task completed in");
      await taskFunction();
      logger.timeEnd("Task completed in");
    }
    catch (error: unknown) {
      logger.error(`Task failed with: ${error}`);
      throw error;
    }
  }

  register.set(taskName, innerTask);
}

export default task;