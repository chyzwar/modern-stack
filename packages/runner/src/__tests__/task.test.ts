import {expect, jest, describe, it} from '@jest/globals';
import task from "../task.js";
import register from "../register.js";
import Logger from "../Logger.js";

jest.mock("../Logger");

describe("task", () => {
  it("should register new task", () => {
    const ls = jest.fn();
    task("ls",  ls);

    expect(register.get("ls")).toBeInstanceOf(Function);
  });

  it("should execute task function when run", async() => {
    const ls = jest.fn();
    task("ls",  ls);  
    
    const run = register.get("ls")!;
    await run();

    expect(Logger.calls).toMatchSnapshot();
    expect(ls).toHaveBeenCalled();
  });
});