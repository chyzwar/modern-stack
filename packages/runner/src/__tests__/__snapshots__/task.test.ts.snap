// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task should execute task function when run 1`] = `
Array [
  Array [
    "Started task",
  ],
  Array [
    "Task completed in",
  ],
  Array [
    "Task completed in",
  ],
]
`;
